### PR TITLE
Overhauled CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by danlevy1**
The CONTRIBUTING.md file has been re-written to be more coherent. Many paragraphs have been turned into numbered steps so that contributors can easily follow it. More sub-sections have been added to improve readability. More information has been added about running Elasticsearch in a Command Line Interface (CLI). More information has also been added about setting up Elasticsearch in Eclipse.

### CODE REVIEW

**Comment by danlevy1** [comment]
I am getting an "All checks have failed" error. I signed the CLA just prior to creating this pull request. It's possible that the CLA has not gone through yet.

**Comment by nik9000** [comment]
I'm sticking this is my review queue for tomorrow morning. It is a lot of changed prose but I'd like to read it. I read the first bit and it looks wonderful.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,208 +1,179 @@
...
+ - Do not worry too much about import order. Try not to change it but there is no need to fight your `IDE` to stop it from doing so.
+
+**Preparing to Submit Changes**
+
+ - To create a distribution from the source, follow these steps:
+	1. Using your favorite Command Line Interface (CLI), `cd` into your cloned Elasticsearch folder
+	2. Run `./gradlew assemble`
+ - The package distribution (Debian and RPM) can be found under ./distribution/packages/(deb|rpm)/build/distributions/
+ - The archive distributions (tar and zip) can be found under ./distribution/archives/(tar|zip)/build/distributions/
+ - Before submitting changes, run the test suite to make sure that nothing is broken: `./gradlew check`
```
Comment (nik9000) [inline CONTRIBUTING.md]: I think we should pull the `check` line to the top because it is the most important thing. I think making the packages is just a thing you might do if you need to test your changes outside of what `gradle run` can do or what you can do with a test. It might be nice to mention that we have tons of testing scenarios in the `qa` directories and that it is perfectly fine to open a PR without a good test if you can't find a good place to add the test. If you make it clear that you *want* to add a test but can't find a spot a reviewer will be more than happy to help.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,208 +1,179 @@
...
+
+**Submitting Changes**
+
+Please follow these steps in order to make sure your code has a chance to make it into the Elasticsearch codebase:
+
+1. Test your changes
```
Comment (nik9000) [inline CONTRIBUTING.md]: I feel like some of this is duplicated from the above section. I think it might be good enough just to say "after you have tested you change" or something like that.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,208 +1,179 @@
...
+		 - Run the test suite to make sure nothing is broken. See the [TESTING](https://github.com/elastic/elasticsearch/blob/master/TESTING.asciidoc) file for help running tests
+2.  Sign the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/) (CLA)
+	- We are not asking you to assign copyright to us, but to give us the right to distribute your code without restriction
+	 - The CLA is used to assure our users of the origin and continuing existence of the code
```
Comment (nik9000) [inline CONTRIBUTING.md]: Can you line up the `-`s?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,208 +1,179 @@
...
+	- You only need to sign the CLA once
+3. Rebase your changes
```
Comment (nik9000) [inline CONTRIBUTING.md]: I'm not sure either of theses things is true any more. To be honest I'm perfectly ok with initial PRs having multiple commits. And I'm quite happy with folks merging `master` into their PR branch before opening rather than rebasing. I think that is good practice because that is what we ask that you do while folks are reviewing the PR.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,208 +1,179 @@
...
+	- Update your local repository with the most recent code from the main Elasticsearch repository and rebase your branch on top of the latest master branch
+	- We prefer your initial changes to be squashed into a single commit. Later, if we ask you to make changes, add them as separate commits. This makes them easier to review.
+	- As a final step before merging, we will either ask you to squash all commits yourself or we will do it for you
```
Comment (nik9000) [inline CONTRIBUTING.md]: This line is totally out of date now. We use the squash feature in github which squashes for us. We merge long running feature branches but not PRs.

Comment (nik9000) [reply CONTRIBUTING.md]: Just drop this line please.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,208 +1,179 @@
...
 4. Submit a pull request
+	- Push your local changes to your forked copy of the repository and submit a [pull request](https://help.github.com/articles/using-pull-requests)
```
Comment (nik9000) [inline CONTRIBUTING.md]: I think we should tell people how to fork before cloning.

I *do* think this is the right place to say that you should push to a named branch on your fork though.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,208 +1,179 @@
...
+		1. **Title**: Sum up the changes that you have made
+		2. **Body**: Provide more details about what your changes do and mention the number of the issue where discussion has taken place, eg "Closes #123."
+
+Once you have completed the above steps, sit back and wait. A discussion about the pull request will most likely be created. If any changes are needed, we would love to work with you to get your pull request merged into Elasticsearch!
+
+Please adhere to the general guideline that you should **never** force push to a **publicly shared** branch. Once you have opened your pull request, you should consider your branch publicly shared. Instead of force pushing, just add incremental commits. This is generally easier on your reviewers. If you need to pick up changes from master, you can merge master into your branch. A reviewer might ask you to rebase a long-running pull request in which case force pushing **is okay** for that request. Note that **squashing** at the end of the review process should also **not** be done. That can be done when the pull request is [integrated via GitHub](https://github.com/blog/2141-squash-your-commits).
```
Comment (nik9000) [inline CONTRIBUTING.md]: I'd reword this to something like "Once you have opened your pull request you should consider its branch **publicly shared** so you should never force push to it. Force pushing makes it much harder to review a PR. Instead of force pushing, add incremental commits. It is fine to merge the base branch into this branch. GitHub will keep the review UI clean."

I think you can drop the section about when it is OK to force push. If the reviewer says it is ok then it is ok. They are declaring bankruptcy on the change tracking process and expect to start over.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,208 +1,179 @@
...
+
+## Contributing as Part of a Class
+
+In general, Elasticsearch is happy to accept contributions that were created as part of a class. However, we strongly advise against making the contribution as part of the class. If you have written code for a class, feel free to submit it.
+
+We ask (_very strongly_) that you do not assign contributing to Elasticsearch as part of a class. If you really want to assign writing code for Elasticsearch as an assignment, please have students contribute to a private clone of Elasticsearch. Opening PRs against the primary Elasticsearch clone must be optional, fully voluntary, not for a grade, and without any deadlines. This is because:
+
+- While the code review process is likely very educations, it can take wildly varying amounts of time depending on who is available, where the change is, and how deep the change is. There is no way to predict how long it will take unless we rush.
```
Comment (nik9000) [inline CONTRIBUTING.md]: s/educations/educational/

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,208 +1,179 @@
...
+
+Please adhere to the general guideline that you should **never** force push to a **publicly shared** branch. Once you have opened your pull request, you should consider your branch publicly shared. Instead of force pushing, just add incremental commits. This is generally easier on your reviewers. If you need to pick up changes from master, you can merge master into your branch. A reviewer might ask you to rebase a long-running pull request in which case force pushing **is okay** for that request. Note that **squashing** at the end of the review process should also **not** be done. That can be done when the pull request is [integrated via GitHub](https://github.com/blog/2141-squash-your-commits).
+
+## Contributing as Part of a Class
+
+In general, Elasticsearch is happy to accept contributions that were created as part of a class. However, we strongly advise against making the contribution as part of the class. If you have written code for a class, feel free to submit it.
```
Comment (nik9000) [inline CONTRIBUTING.md]: I wrote this section originally and now that I reread it I'm not sure it makes much sense. But I'm not sure that your changes are better.

The gist is:
1. For the sake of all that is holy do not make someone's grade contingent on getting something merged to our repo.
2. If you feel like writing code for Elasticsearch as part of a class, we'll review it and merge it, but in our own time.
3. Do not tell a bunch of students to go and cookie lick issues and then make a half hearted attempt at solving them. I don't care how good you think your students are, we can't handle 30 new contributors in the span of a week. We certainly can't handle them if they aren't doing this of their own free will, excited to make changes and get them merged.

Most of this comes down to time and resources. I'd love new contributors and am fine with them being in university. But we don't operate on academic timelines. Sometimes we're faster and sometimes we're slower. And I'm really worried that the incentives for students to contribute are messed up. I've had folks say they only plan to open the PR but not to do the work suggested by the reviewers to merge it because that is all their class wanted from them....

**Review by nik9000 — CHANGES_REQUESTED** [review]
I left a bunch of suggestions. A few more that don't make it:
1. Rewriting everything in this document is fine with me. I think it deserves it rather than the patchwork patching we've been doing to it.
2. I think you should break your lines somewhere. I tend to go with 80 or 100 or something. That'll make it easier to review the change because I can leave notes closer to the text you are changing.
3. It think this might take a few rounds of review. That is fine with me if it is fine with you.
4. I mentioned adding a section for making documentation changes. I can add that section in a followup to your change if you'd like.
5. Can you merge master into this? There are merge conflicts.

**Comment by colings86** [comment]
@danlevy1 are you still interested in working on this?

**Comment by nik9000** [comment]
@javanna yes and no. I'll add reworking it myself to my TODO list but close this.
